CotEditor/Sources/Script.swift

```diff
@@ -28,68 +28,163 @@
 
 import Cocoa
 
-class Script {
+enum ScriptingEventType: String {
     
-    let url: URL
-    let name: String
+    case documentOpened = "document opened"
+    case documentSaved = "document saved"
     
     
-    
-    // MARK: -
-    // MARK: Lifecycle
-    
-    init(url: URL, name: String) {
+    var eventID: AEEventID {
         
-        self.url = url
-        self.name = name
+        switch self {
+        case .documentOpened: return AEEventID(code: "edod")
+        case .documentSaved: return AEEventID(code: "edsd")
+        }
     }
     
+}
+
+
+
+enum ScriptingFileType {
+    
+    case appleScript
+    case shellScript
+    case unknown
```

Comment:
Aha, ok, right. It can be an unknown file. Then, how about:

1. Changing `unknown` to `folder`.
2. Make ScriptDescriptor's `init(at url: URL)` to `init?(at url: URL)` so that it can be nil if the file at given URL is an unknown file.